### DISCUSSION TITLE
无线表格的推理cpu 连续跑10次平均将进0.93，不只500ms呢,GPU 1s

### DISCUSSION THREAD

**Discussion by cqray1990** [RapidAI/TableStructureRec#54] [Q&A]
![lineless_table_recognition](https://github.com/user-attachments/assets/01b34d9b-5854-4410-98a1-03ede1559693)

**Comment by Joker1212** [comment]
> ![lineless_table_recognition](https://private-user-images.githubusercontent.com/32585434/379997023-01b34d9b-5854-4410-98a1-03ede1559693.jpg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk4MjY3MjYsIm5iZiI6MTcyOTgyNjQyNiwicGF0aCI6Ii8zMjU4NTQzNC8zNzk5OTcwMjMtMDFiMzRkOWItNTg1NC00NDEwLTk4YTEtMDNlZGUxNTU5NjkzLmpwZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI1VDAzMjAyNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM0YWJmZTRlYTM5YzY5YWJiNzBiNzFjNWM1N2VkMTNkNmNmNWFhMDk4ZTJjMmNiYjc2NzllNTExMDgxYjQ5NzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.IgiCtL4HcgXAK3apKx3GO1MPeddcdkWnE-ar3nKNKzI)

emmm,请问是什么cpu呢，我这边是mac m1测的，gpu的话理论上还是onnx模型本身结构支持不够好，有一个简单的办法是进行量化，或者可以尝试slanet-plus当做无线表格模型哈，是因为cls模型精度不够才没有把它当无线模型用

**Reply by Joker1212** [reply]
量化的话，直接用onnx的量化工具进行动态量化就可以了，输入输出保持不变

**Reply by cqray1990** [reply]
> 量化的话，直接用onnx的量化工具进行动态量化就可以了，输入输出保持不变

量化多少位的呢？cpu ：Intel® Core™ i7-8700 CPU @ 3.20GHz × 12

**Reply by Joker1212** [reply]
就是默认的qint8就行，或者quint8也行

**Reply by Joker1212** [reply]
可以参考这个，稍微改一下就行
```python
from pathlib import Path
import onnx
from onnxruntime.quantization import quantize_dynamic, QuantType, quant_pre_process
def quantize_model(root_dir_str, model_dir, pre_fix):

    original_model_path = f"{pre_fix}.onnx"
    quantized_model_path = f"{pre_fix}_quantized.onnx"
    # quantized_model_path = original_model_path
    original_model_path = f"{root_dir_str}/{model_dir}/{original_model_path}"
    quantized_model_path = f"{root_dir_str}/{model_dir}/{quantized_model_path}"
    quant_pre_process(original_model_path, quantized_model_path, auto_merge=True)
    # 进行动态量化
    quantize_dynamic(
        model_input=original_model_path,
        model_output=quantized_model_path,
        weight_type=QuantType.QUInt8
    )

    # 验证量化后的模型
    quantized_model = onnx.load(quantized_model_path)
    onnx.checker.check_model(quantized_mo
```

**Reply by cqray1990** [reply]
> from pathlib import Path import onnx from onnxruntime.quantization import quantize_dynamic, QuantType, quant_pre_process def quantize_model(root_dir_str, model_dir, pre_fix):
> 
> ```
> original_model_path = f"{pre_fix}.onnx"
> quantized_model_path = f"{pre_fix}_quantized.onnx"
> # quantized_model_path = original_model_path
> original_model_path = f"{root_dir_str}/{model_dir}/{original_model_path}"
> quantized_model_path = f"{root_dir_str}/{model_dir}/{quantized_model_path}"
> quant_pre_process(original_model_path, quantized_model_path, auto_merge=True)
> # 进行动态量化
> quantize_dynamic(
>     model_input=quantized_model_path,
>     model_output=quantized_model_path,
>     weight_type=QuantType.QUInt8
> )
> 
> # 验证量化后的模型
> quantized_model = onnx.load(quantized_model_path)
> onnx.checker.check_model(quantized_mo
> ```

量化模型肯定影响效果，哈哈

**Reply by Joker1212** [reply]
理论上影响很小其实，你可以量化了试试你的场景受影响的情况

**Reply by cqray1990** [reply]
> 理论上影响很小其实，你可以量化了试试你的场景受影响的情况

合合TextIn的无线表格效果很通用，不知道他们怎么做的

**Reply by Joker1212** [reply]
好像没有公开他们的实现，只知道他们也是区分了有线和无线场景分别训练的模型，看起来可能是transformer的序列
